crates/ruff_python_semantic/src/definition.rs

```diff
@@ -113,13 +123,76 @@ impl<'a> Definitions<'a> {
         next_id
     }
 
-    /// Iterate over all definitions in a program, with their visibilities.
-    pub fn iter(&self) -> DefinitionsIter {
-        DefinitionsIter {
-            inner: self.0.iter(),
-            definitions: Vec::with_capacity(self.0.len()),
-            visibilities: Vec::with_capacity(self.0.len()),
+    /// Resolve the visibility of each definition in the collection.
+    pub fn resolve(self, exports: Option<&[&str]>) -> ContextualizedDefinitions<'a> {
```

Comment:
I changed this from a stateful iterator to a separate `ContextualizedDefinitions` struct. (We can't use an iterator because of the lifetimes: using `exports` is an immutable borrow of `Context`, and we can't be borrowing `Context` here.)

Using all these structs feels kind of clunky, open to feedback.

Reply:
One option could be to simply clone `Exports.names`. It should be relatively cheap because it only runs once and should only contains few entries.